cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go

```diff
@@ -246,3 +249,33 @@ func pathForKey(pkiPath, name string) string {
 func pathForPublicKey(pkiPath, name string) string {
 	return filepath.Join(pkiPath, fmt.Sprintf("%s.pub", name))
 }
+
+func BackupAPIServerCertAndKey() error {
+	certAndKeyPath := map[string]string{
+		constants.APIServerCertName: filepath.Join(kubeadmapiext.DefaultCertificatesDir, constants.APIServerCertName),
+		constants.APIServerKeyName:  filepath.Join(kubeadmapiext.DefaultCertificatesDir, constants.APIServerKeyName),
+	}
+
+	certAndKeyBackupPath := map[string]string{
+		constants.APIServerCertName: filepath.Join(kubeadmapiext.DefaultCertificatesBackupDir, constants.APIServerCertName),
+		constants.APIServerKeyName:  filepath.Join(kubeadmapiext.DefaultCertificatesBackupDir, constants.APIServerKeyName),
+	}
+
+	rollbackCertAndKey := func(paths map[string]string, backupPaths map[string]string, originalErr error) error {
+		errs := []error{originalErr}
+		for name, backupPath := range backupPaths {
+			if err := os.Rename(backupPath, paths[name]); err != nil {
+				errs = append(errs, err)
+			}
+		}
+		return fmt.Errorf("couldn't backup apiserver cert and key! Got errors: %v", errors.NewAggregate(errs))
+	}
+
+	for name, path := range certAndKeyPath {
```

Comment:
I would simplify this to something like this:
```go
func MoveFiles(files map[string]string) error {
    filesToRecover := map[string]string{}
    for from, to := range files {
          if err := os.Rename(from, to); err != nil {
                 return rollbackCertAndKey(from, to, err)
          }
          filesToRecover[from] = to
    }
    return nil
}
```
and modify the rollback func accordingly.
Then pass in the right map in a func in a new file called `postupgrade_v18_v19.go`:
```go
func BackupAPIServerCertAndKey(certDir string) error {
    subDir := "expired"
    filesToMove := map[string]string{
         filepath.Join(certDir, constants.APIServerCertName): filepath.Join(certDir, subDir, constants.APIServerCertName),
         filepath.Join(certDir, constants.APIServerKeyName): filepath.Join(certDir, subDir, constants.APIServerKeyName),
    }
    ...
}
```